modules/uhttp/filters.go

```diff
@@ -48,76 +49,71 @@ func (f FilterFunc) Apply(ctx fx.Context, w http.ResponseWriter, r *http.Request
-		}
-		span := opentracing.GlobalTracer().StartSpan(operationName, ext.RPCServerOption(spanCtx))
-		ext.HTTPUrl.Set(span, r.URL.String())
-		defer func() {
-			span.Finish()
-			fxctx.Logger().Debug("Span finished")
-		}()
-
-		fxctx = fxctx.WithContextAwareLogger(span)
-		fxctx = &fxcontext.Context{
-			Context: opentracing.ContextWithSpan(fxctx, span),
-		}
-		r = r.WithContext(fxctx)
-		next.ServeHTTP(fxctx, w, r)
+func (f tracingServerFilter) Apply(ctx fx.Context, w http.ResponseWriter, r *http.Request, next Handler) {
+	operationName := r.Method
+	carrier := opentracing.HTTPHeadersCarrier(r.Header)
+	spanCtx, err := opentracing.GlobalTracer().Extract(opentracing.HTTPHeaders, carrier)
+	if err != nil && err != opentracing.ErrSpanContextNotFound {
+		ctx.Logger().Info("Malformed inbound tracing context: ", "error", err.Error())
 	}
+	span := opentracing.GlobalTracer().StartSpan(operationName, ext.RPCServerOption(spanCtx))
+	ext.HTTPUrl.Set(span, r.URL.String())
+	defer span.Finish()
+
+	ctx = ctx.(*fxcontext.Context).WithContextAwareLogger(span)
+	ctx = &fxcontext.Context{
+		Context: opentracing.ContextWithSpan(ctx, span),
+	}
+	r = r.WithContext(ctx)
+	next.ServeHTTP(ctx, w, r)
 }
 
 // authorizationFilter authorizes services based on configuration
-func authorizationFilter(host service.Host) FilterFunc {
-	return func(ctx context.Context, w http.ResponseWriter, r *http.Request, next Handler) {
-		fxctx := &fxcontext.Context{
-			Context: ctx,
-		}
+type authorizationFilter struct {
+	authClient auth.Client
+	scope      tally.Scope
+}
 
-		if err := host.AuthClient().Authorize(fxctx); err != nil {
-			host.Metrics().SubScope("http").SubScope("auth").Counter("fail").Inc(1)
-			fxctx.Logger().Error(auth.ErrAuthorization, "error", err)
-			w.WriteHeader(http.StatusUnauthorized)
-			fmt.Fprintf(w, "Unauthorized access: %+v", err)
-			return
-		}
-		next.ServeHTTP(fxctx, w, r)
+func (f authorizationFilter) Apply(ctx fx.Context, w http.ResponseWriter, r *http.Request, next Handler) {
+	if err := f.authClient.Authorize(ctx); err != nil {
+		f.scope.Counter("auth.fail").Inc(1)
+		ctx.Logger().Error(auth.ErrAuthorization, "error", err)
+		w.WriteHeader(http.StatusUnauthorized)
+		fmt.Fprintf(w, "Unauthorized access: %+v", err)
+		return
 	}
+	next.ServeHTTP(ctx, w, r)
 }
 
 // panicFilter handles any panics and return an error
 // panic filter should be added at the end of filter chain to catch panics
-func panicFilter(host service.Host) FilterFunc {
-	return func(ctx context.Context, w http.ResponseWriter, r *http.Request, next Handler) {
-		fxctx := &fxcontext.Context{
-			Context: ctx,
+type panicFilter struct {
+	scope tally.Scope
+}
+
+func (f panicFilter) Apply(ctx fx.Context, w http.ResponseWriter, r *http.Request, next Handler) {
+	defer func() {
+		if err := recover(); err != nil {
+			ctx.Logger().Error("Panic recovered serving request", "error", err, "url", r.URL)
+			f.scope.Counter("panic").Inc(1)
```

Comment:
i don't recall exactly how these work, but i thought the general guidance was to avoid instantiating counters on-demand, and instead frontload the allocation